### PR TITLE
benchmarking automation pallet on init

### PR DESCRIPTION
**Pull request by justinzhou93**

### CODE REVIEW

**Comment by justinzhou93** [comment]
tested functionality locally, still looks good

File: pallets/automation-time/src/tests.rs [hunk 1]
```diff
@@ -511,7 +511,7 @@ fn trigger_tasks_completes_some_tasks() {
 		);
 		LastTimeSlot::<Test>::put(LAST_BLOCK_TIME);
 
-		AutomationTime::trigger_tasks(90_000);
+		AutomationTime::trigger_tasks(80_000);
```
Comment (justinzhou93) [inline pallets/automation-time/src/tests.rs:514]: i think the loop before checked for 60k, now is only checking for 10k, so this is why we need to lower the total weight, else everything runs

File: pallets/automation-time/src/weights.rs [hunk 2]
```diff
@@ -52,15 +70,15 @@ impl<T: frame_system::Config> WeightInfo for AutomationWeight<T> {
...
+		(15_000_000 as Weight)
 			.saturating_add(T::DbWeight::get().reads(3 as Weight))
 			.saturating_add(T::DbWeight::get().writes(2 as Weight))
 	}
 	// Storage: Timestamp Now (r:1 w:0)
 	// Storage: AutomationTime Tasks (r:1 w:1)
 	// Storage: AutomationTime ScheduledTasks (r:1 w:1)
 	fn schedule_notify_task_full() -> Weight {
-		(15_000_000 as Weight)
+		(31_000_000 as Weight)
```
Comment (justinzhou93) [inline pallets/automation-time/src/weights.rs:81]: idk why this changed so much

File: rust-toolchain.toml [hunk 3]
```diff
@@ -1,3 +1,3 @@
 [toolchain]
-channel = "nightly-2021-12-15"
+channel = "stable-aarch64-apple-darwin"
```
Comment (justinzhou93) [inline rust-toolchain.toml:2]: is this good to save to github too?

File: runtime/neumann/src/lib.rs [hunk 4]
```diff
@@ -720,7 +720,7 @@ impl pallet_treasury::Config for Runtime {
 }
 
 parameter_types! {
-	pub const MaxTasksPerSlot: u32 = 2;
+	pub const MaxTasksPerSlot: u32 = 100;
```
Comment (justinzhou93) [inline runtime/neumann/src/lib.rs:723]: can i actually change this?

File: pallets/automation-time/src/lib.rs [hunk 5]
```diff
@@ -453,28 +453,28 @@ pub mod pallet {
 			weight_left -= update_weight;
 
 			// need to calculate the weight of running just 1 task below.
-			if weight_left < 60_000 {
+			if weight_left < <T as Config>::WeightInfo::run_tasks_many_found(1) {
```
Comment (justinzhou93) [inline pallets/automation-time/src/lib.rs:456]: not sure this value should actually be:
```suggestion
			if weight_left < (<T as Config>::WeightInfo::run_tasks_many_found(1) + <T as Config>::WeightInfo::read() + <T as Config>::WeightInfo::write()) {
```

File: pallets/automation-time/src/lib.rs [hunk 6]
```diff
@@ -453,28 +453,28 @@ pub mod pallet {
...
-				weight_left -= 10_000;
+				weight_left -= <T as Config>::WeightInfo::write();
 				let (tasks_left, new_weight_left) = Self::run_tasks(task_queue, weight_left);
 				TaskQueue::<T>::put(tasks_left);
 				weight_left = new_weight_left;
 			}
 
 			// if there is weight left we need to handled the missed tasks
-			if weight_left >= 60_000 {
+			if weight_left >= <T as Config>::WeightInfo::run_missed_tasks_many_found(1) {
```
Comment (justinzhou93) [inline pallets/automation-time/src/lib.rs:472]: same as above, should i add a read and a write to this calculation?

File: pallets/automation-time/src/lib.rs [hunk 7]
```diff
@@ -588,15 +586,15 @@ pub mod pallet {
...
 					},
 				};
 
 				// need to calculate the look cost minus the action
-				weight_left = weight_left - action_weight - 10_000;
+				weight_left = weight_left - action_weight;
 
 				// need to calculate the max cost of the loop
-				if weight_left < 20_000 {
+				if weight_left < <T as Config>::WeightInfo::run_tasks_many_found(1) {
```
Comment (justinzhou93) [inline pallets/automation-time/src/lib.rs:597]: just to double check my sanity here, max of loop should just be run_tasks_many_found, right?

File: pallets/automation-time/src/lib.rs [hunk 8]
```diff
@@ -612,37 +610,34 @@ pub mod pallet {
...
 					},
 				};
 
 				// need to calculate the look cost minus the action
-				weight_left = weight_left - action_weight - 10_000;
+				weight_left = weight_left - action_weight;
 
 				// need to calculate the max cost of the loop
-				if weight_left < 20_000 {
+				if weight_left < <T as Config>::WeightInfo::run_missed_tasks_many_found(1) {
```
Comment (justinzhou93) [inline pallets/automation-time/src/lib.rs:640]: just to double check my sanity here, max of loop should just be run_missed_tasks_many_found, right?

File: pallets/automation-time/src/lib.rs [hunk 9]
```diff
@@ -654,14 +649,35 @@ pub mod pallet {
 			}
 		}
 
+		pub fn test_missing_tasks_remove_events(
```
Comment (justinzhou93) [inline pallets/automation-time/src/lib.rs:652]: should i actually just use this function for benchmarking? or should i refactor the existing code to use this?

**Review by justinzhou93 — COMMENTED** [review]
oh jeez, i realized i did everything with notify event. i should prob swap out all of my benchmarking with transfer instead, right?

File: pallets/automation-time/src/benchmarking.rs [hunk 10]
```diff
@@ -45,6 +45,22 @@ fn schedule_tasks<T: Config>(owner: T::AccountId, time: u64, count: u32) -> T::H
 	task_id
 }
 
+fn schedule_transfer_tasks<T: Config>(owner: T::AccountId, time: u64, count: u32, recipient: T::AccountId, amount: BalanceOf<T>) -> T::Hash {
```
Comment (rhuttman) [inline pallets/automation-time/src/benchmarking.rs:48]: no real need to have `recipient` and `amount` as params. Can just hardcode and say what they are in a fcn comment

File: pallets/automation-time/src/benchmarking.rs [hunk 11]
```diff
@@ -45,6 +45,22 @@ fn schedule_tasks<T: Config>(owner: T::AccountId, time: u64, count: u32) -> T::H
...
+	let time_moment: u32 = time.try_into().unwrap();
+	<pallet_timestamp::Pallet<T>>::set_timestamp(time_moment.into());
+	let mut task_id: T::Hash = T::Hash::default();
+	let converted_count: u8 = count.try_into().unwrap();
```
Comment (rhuttman) [inline pallets/automation-time/src/benchmarking.rs:52]: should prop use saturate into. I didn't know about it when i made the fcn above

File: pallets/automation-time/src/benchmarking.rs [hunk 12]
```diff
@@ -138,4 +156,205 @@ benchmarks! {
...
+
+		let amount_transferred: BalanceOf<T> = T::ExistentialDeposit::get().saturating_mul(ED_MULTIPLIER.into());
+		let task_id: T::Hash = schedule_transfer_tasks::<T>(caller.clone(), time, 1, recipient.clone(), amount_transferred.clone());
+	}: { AutomationTime::<T>::run_native_transfer_task(caller, recipient, amount_transferred, task_id) }
+
+	/*
+	* This section is to test run_missed_tasks.
+	* run_missed_tasks_none: measuring base overhead of run_missed_tasks
+	* run_missed_tasks_many_found: measuring many tasks for linear progression
+	* run_missed_tasks_many_missing: measuring number of tasks if not found in queue, to find slope of growth
```
Comment (rhuttman) [inline pallets/automation-time/src/benchmarking.rs:180]: isn't this also a linear progression?

File: pallets/automation-time/src/benchmarking.rs [hunk 13]
```diff
@@ -138,4 +156,205 @@ benchmarks! {
...
+			let task_id = AutomationTime::<T>::generate_task_id(caller.clone(), provided_id);
+			task_ids.push(task_id)
+		}
+	}: { AutomationTime::<T>::run_missed_tasks(task_ids, weight_left) }
+
+	/*
+	* This section is to test run_tasks.
+	* run_tasks_none: tests base framework of run_tasks
+	* run_tasks_many_found: tests many tasks for linear progression
+	* run_tasks_many_missing: tests number of tasks if not found in queue, to find slope of growth
```
Comment (rhuttman) [inline pallets/automation-time/src/benchmarking.rs:240]: is this not linear?

File: pallets/automation-time/src/benchmarking.rs [hunk 14]
```diff
@@ -138,4 +156,205 @@ benchmarks! {
...
+
+		for i in 0..v {
+			let provided_id: Vec<u8> = vec![i.saturated_into::<u8>()];
+			let task_id = AutomationTime::<T>::generate_task_id(caller.clone(), provided_id);
+			task_ids.push(task_id)
+		}
+	}: { AutomationTime::<T>::run_tasks(task_ids, weight_left) }
+
+	/*
+	* This section is to test update_task_queue.
```
Comment (rhuttman) [inline pallets/automation-time/src/benchmarking.rs:283]: prob worth calling out this only tests for 1 missed time slot. 

Can we use this to infer the weight if there is no missed slot?

File: pallets/automation-time/src/benchmarking.rs [hunk 15]
```diff
@@ -138,4 +156,205 @@ benchmarks! {
...
+			let provided_id: Vec<u8> = vec![i.saturated_into::<u8>()];
+			let task_id = AutomationTime::<T>::schedule_task(caller.clone(), provided_id.clone(), time.into()).unwrap();
+			let task = Task::<T>::create_event_task(caller.clone(), provided_id, time.into(), vec![4, 5, 6]);
+			<Tasks<T>>::insert(task_id, task);
+			task_ids.push(task_id)
+		}
+	}: { AutomationTime::<T>::test_missing_tasks_remove_events(task_ids) }
+
+	/*
+	* Utils: read and write
```
Comment (rhuttman) [inline pallets/automation-time/src/benchmarking.rs:354]: I don't get this section

**Review by rhuttman — APPROVED** [review]

**Pull request merged**